tools/sim/bridge/carla.py

```diff
@@ -0,0 +1,171 @@
+    self.imu = world.spawn_actor(imu_bp, transform, attach_to=vehicle)
+
+    gps_bp = blueprint_library.find('sensor.other.gnss')
+    self.gps = world.spawn_actor(gps_bp, transform, attach_to=vehicle)
+    self.params.put_bool("UbloxAvailable", True)
+
+    self.carla_objects = [self.imu, self.gps, self.road_camera, self.road_wide_camera]
+
+  def close(self):
+    for s in self.carla_objects:
+      if s is not None:
+        try:
+          s.destroy()
+        except Exception as e:
+          print("Failed to destroy carla object", e)
+
+  def carla_image_to_rgb(self, image):
+    rgb = np.frombuffer(image.raw_data, dtype=np.dtype("uint8"))
+    rgb = np.reshape(rgb, (H, W, 4))
+    return np.ascontiguousarray(rgb[:, :, [0, 1, 2]])
+
+  def cam_callback_road(self, image):
+    with self.image_lock:
+      self.road_image = self.carla_image_to_rgb(image)
+
+  def cam_callback_wide_road(self, image):
+    with self.image_lock:
+      self.wide_road_image = self.carla_image_to_rgb(image)
+
+  def apply_controls(self, steer_angle, throttle_out, brake_out):
+    self.vc.throttle = throttle_out
+
+    steer_carla = steer_angle * -1 / (self.max_steer_angle * self.steer_ratio)
+    steer_carla = np.clip(steer_carla, -1, 1)
+
+    self.vc.steer = steer_carla
+    self.vc.brake = brake_out
+    self.vehicle.apply_control(self.vc)
+
+  def read_sensors(self, simulator_state: SimulatorState):
+    simulator_state.imu.bearing = self.imu.get_transform().rotation.yaw
+
+    simulator_state.imu.accelerometer = vec3(
+      self.imu.get_acceleration().x,
+      self.imu.get_acceleration().y,
+      self.imu.get_acceleration().z
+    )
+
+    simulator_state.imu.gyroscope = vec3(
+      self.imu.get_angular_velocity().x,
+      self.imu.get_angular_velocity().y,
+      self.imu.get_angular_velocity().z
+    )
+
+    simulator_state.gps.from_xy([self.vehicle.get_location().x, self.vehicle.get_location().y])
+
+    simulator_state.velocity = self.vehicle.get_velocity()
+    simulator_state.valid = True
+    simulator_state.steering_angle = self.vc.steer * self.max_steer_angle
+
+  def read_cameras(self):
+    pass # cameras are read within a callback for carla
+
+  def tick(self):
+    self.world.tick()
+
+
+class CarlaBridge(SimulatorBridge):
+  TICKS_PER_FRAME = 5
+
+  def __init__(self, arguments):
+    super().__init__(arguments)
+    self.host = arguments.host
+    self.port = arguments.port
+    self.town = arguments.town
+    self.num_selected_spawn_point = arguments.num_selected_spawn_point
+
+  def spawn_world(self):
+    import carla  # pylint: disable=import-error
+    def connect_carla_client(host: str, port: int):
+      client = carla.Client(host, port)
+      client.set_timeout(5)
+      return client
+
+    client = connect_carla_client(self.host, self.port)
+    world = client.load_world(self.town)
+
+    settings = world.get_settings()
+    #settings.synchronous_mode = True  # Enables synchronous mode
```

Comment:
is this dead?

Reply:
runs much better without it, removed the commented line